src/libraries/System.Net.Security/src/System/Net/Security/NetSecurityTelemetry.cs

```diff
@@ -0,0 +1,288 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Security.Authentication;
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace System.Net.Security
+{
+    [EventSource(Name = "System.Net.Security")]
+    internal sealed class NetSecurityTelemetry : EventSource
+    {
+        public static readonly NetSecurityTelemetry Log = new NetSecurityTelemetry();
+
+        private IncrementingPollingCounter? _tlsHandshakeRateCounter;
+        private PollingCounter? _totalTlsHandshakesCounter;
+        private PollingCounter? _currentTlsHandshakesCounter;
+        private PollingCounter? _failedTlsHandshakesCounter;
+        private PollingCounter? _connectionsOpenTls10Counter;
```

Comment:
Do we need to keep this per protocol version? That is going hard to maintain IMHO. (and not covering old SSL)

Reply:
As far as I know, counters can't carry other data, so I don't know if this can be simplified

Reply:
The key part is if ween stats per protocol. If we do, we have no option IMHO. In that case it may be worth of adding deprecating Ssl2/3 for completeness. And perhaps add some comment somewhere to update this when next TLS is out. I did not pay much attention. but do we could http1 and http2 separate? From SslStream prospective there is really not that much difference.

Reply:
You're suggesting we add counters for Ssl2/3 as well even tho they're deprecated?

> And perhaps add some comment somewhere to update this when next TLS is out.

Will do, we can also add asserts for that.

> but do we could http1 and http2 separate? From SslStream prospective there is really not that much difference.

Do you mean have separate counters for open connections depending on whether they are used for Http1/Http2?
Afaik SslStream doesn't have any such data right?

Reply:
no it does not. I was wondering if HTTP counters are also separated by protocol versions.
I was wondering why somebody would want to track handshakes by protocol version. 
The answer may be monitoring of very old or very new versions to manage deprecation and adoption.
If we don't add deprecated Ssl, than the counters will not add-up to total number for handshakes. 
That may be fine if we document it or if we add some "other" bucket.

Reply:
HTTP counters aren't split by protocol version, the RequestStart **event** does log the version however.

Reply:
> If we don't add deprecated Ssl, than the counters will not add-up to total number for handshakes.

There is currently no counter for "total-opened-sessions" (though we can add it as it seems useful), there are only per-protocol ones (excluding Ssl2, 3).

Reply:
> And perhaps add some comment somewhere to update this when next TLS is out.

I added an assert that should catch that

> There is currently no counter for "total-opened-sessions"

I added `all-sessions-open` and `all-handshake-duration` counters.

My understanding is that in most cases, a user would have to manually opt-in to using Ssl2/Ssl3, so I would be okay with not adding counters for those. They are still tracked in the `all-*` counters, so documenting it seems better.